kind.go

```diff
@@ -45,48 +47,204 @@ func (m Maturity) String() string {
 type Kind interface {
-	// Props returns a [kindsys.SomeKindProps], representing the properties
+	// Name returns the kind's name, as specified in the name field of the kind definition.
+	//
+	// Note that this is the capitalized name of the kind. For other names and
+	// common kind properties, see [Props.CommonProperties].
+	Name() string
+
+	// MachineName returns the kind's machine name, as specified in the machineName
+	// field of the kind definition.
+	MachineName() string
+
+	// Maturity indicates the maturity of this kind, one of the enum of values we
+	// accept in the maturity field of the kind definition.
+	Maturity() Maturity
+
+	// Props returns a [SomeKindProps], representing the properties
 	// of the kind as declared in the .cue source. The underlying type is
 	// determined by the category of kind.
-	//
-	// This method is largely for convenience, as all actual kind categories are
-	// expected to implement one of the other interfaces, each of which contain
-	// a Def() method through which these same properties are accessible.
 	Props() SomeKindProperties
 
-	// TODO docs
-	Lineage() thema.Lineage
+	// CurrentVersion returns the version number of the schema that is considered
+	// the 'current' version, usually the latest version. When initializing object
+	// instances of this Kind, the current version is used by default.
+	CurrentVersion() thema.SyntacticVersion
 
-	// TODO remove, unnecessary with Props()
-	Name() string
+	// Lineage returns the kind's [thema.Lineage]. The lineage contains the full
+	// history of object schemas associated with the kind.
+	//
+	// TODO separate this onto an optional, additional interface
+	Lineage() thema.Lineage
+}
 
-	// TODO remove, unnecessary with Props()
-	MachineName() string
+// ResourceKind represents a kind that defines a root object, or Kubernetish resource.
+//
+// TODO this is a temporary intermediate while we combine [Core] and [Custom]. This name will probably go away.
+type ResourceKind interface {
+	Kind
 
-	// TODO remove, unnecessary with Props()
-	Maturity() Maturity // TODO unclear if we want maturity for raw kinds
+	// Validate takes a []byte representing an object instance of this kind and
+	// checks that it is a valid instance of at least one schema defined in the
+	// kind.
+	//
+	// A decoder must be provided that knows how to decode the []byte into an
+	// intermediate form. At minimum, the right decoder must be chosen for the
+	// format - for example, JSON vs YAML. For resource kinds, a decoder must
+	// also know how to transform the input from a Kubernetes resource object
+	// shape to Grafana's object shape. See [github.com/grafana/kindsys/encoding].
+	Validate(b []byte, codec Decoder) error
+
+	// FromBytes takes a []byte and a decoder, validates it against schema, and
+	// if validation is successful, unmarshals it into an UnstructuredResource.
+	FromBytes(b []byte, codec Decoder) (*UnstructuredResource, error)
```

Comment:
> Can we think of a good justification to require the decoder in the method, which then requires that implementers call codec.Decode at the start to produce encoding.GrafanaShapeBytes, rather than just having the method accept encoding.GrafanaShapeBytes, and 

Even though the current implementation of the decoders is ugly, my main thinking here is that it's the same pattern that made vmuxers work tidily. i think that's because there's a basic set of responsibilities here:

* Getting some bytes from somewhere, and knowing what shape to expect is in them (caller)
* Transforming bytes (decoder) of some known shape (caller) into an intermediate form
* Transforming intermediate form into known Go type (method impl)

That said, i'm not really settled on this - just settled _enough_ that i think it's fine for this initial PR.

Reply:
Sounds good enough for now, I think this whole interface will likely have several evolutions before its final form.